`azurerm_api_management_backend` - support new property `circuit_breaker_rule`

sinbai opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

Support new property `circuit_breaker_rule`.

Swagger: https://github.com/Azure/azure-rest-api-specs/blob/551674fef42784a6d62d92149d637a5d450f5128/specification/apimanagement/resource-manager/Microsoft.ApiManagement/stable/2024-05-01/definitions.json#L1979


Note: This is a partial fix for #24594, as load-balanced pools are not supported due to API limitations. Support will be added once a new API version is released that fixes the reported issue.

## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->

PASS: TestAccApiManagementBackend_circuitBreakerRule (538.33s)

<img width="493" height="420" alt="image" src="https://github.com/user-attachments/assets/89a62121-06a0-4653-ac4e-b5c642e6101b" />


## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_api_management_backend` - support for the `circuit_breaker_rule` property [GH-00000]


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [ ] Bug Fix
- [x] New Feature (ie adding a service, resource, or data source)
- [x] Enhancement
- [ ] Breaking Change


## Related Issue(s)
Fixes #24594

<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->

## Rollback Plan

If a change needs to be reverted, we will publish an updated version of the provider.

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

magodo reviewed (changes requested):
Thanks for this PR!
I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

sinbai commented on internal/services/apimanagement/api_management_backend_resource.go:
```diff
@@ -531,6 +611,142 @@ func expandApiManagementBackendTls(input []interface{}) *backend.BackendTlsPrope
+
+	if name := v["name"]; name != nil && name.(string) != "" {
+		rule.Name = pointer.To(name.(string))
+	}
```
Yes, fixed.

sinbai commented on internal/services/apimanagement/api_management_backend_resource.go:
```diff
@@ -531,6 +611,142 @@ func expandApiManagementBackendTls(input []interface{}) *backend.BackendTlsPrope
+	}
+
+	if hasCondition {
+		rule.FailureCondition = pointer.To(condition)
```
I seem to recall receiving similar feedback from HC about flattening when there are many levels. However, after reviewing the schema-design-considerations.md document, I couldn't find such a requirement. So, I updated the code to keep the Swagger structure consistent. Thanks a lot for your feedback!

sinbai commented on internal/services/apimanagement/api_management_backend_resource.go:
```diff
@@ -531,6 +611,142 @@ func expandApiManagementBackendTls(input []interface{}) *backend.BackendTlsPrope
+		result["accept_retry_after_enabled"] = pointer.From(rule.AcceptRetryAfter)
+
+		if rule.FailureCondition != nil {
+			condition := rule.FailureCondition
```
Fixed.

sinbai commented:
> Thanks for this PR! I've taken a look through and left some comments inline, but this is mostly looking good to me 👍

Hi @magodo I've addressed the comments - would you mind taking another look when you have a chance?

magodo commented:
LGTM!

<img width="418" height="181" alt="image" src="https://github.com/user-attachments/assets/10386d3c-c54d-4623-8b05-c7edd1b9ff98" />

whitelogicsolutions commented:
@WodansSon please review it, this will be a useful feature. Thanks @sinbai for adding.
